### ISSUE TITLE
Null Value Payload Indexing

### ISSUE THREAD

**Issue by esteininger** [qdrant/qdrant#5342]
**Is your feature request related to a problem? Please describe.**
When querying for null values in payload fields (e.g., `metadata.asset_url == null`), the current implementation requires reading directly from the payload, resulting in poor performance. This is particularly problematic when the null check is a frequent operation that appears in most queries. Currently, the only workaround is to convert null values to booleans, which creates overhead during data insertion.

Example of the current frustration:
- Cannot efficiently query for records where a specific field exists but contains null
- `is_empty` checks presence in indices but doesn't distinguish between null and empty fields
- Direct null comparison (`eq null`) requires full payload scan, making it prohibitively expensive for frequent queries

**Describe the solution you'd like**
Add native support for efficient null value querying in payload fields, specifically:
1. Implement index optimization for null value lookups
2. Add ability to distinguish between null values and missing fields in indices
3. Make `eq null` queries as performant as other indexed field queries

This would allow operations like `metadata.asset_url == null` to execute efficiently without requiring payload scans or type conversion workarounds.

**Describe alternatives you've considered**
1. Using boolean field conversion:
   - Converting null values to boolean flags during insertion
   - Creates overhead during data insertion
   - Requires maintaining additional fields
   
2. Using `is_empty`:
   - Not suitable as it doesn't distinguish between null and missing fields
   - Doesn't solve the specific use case

3. Using direct null comparison:
   - Currently too slow for frequent queries
   - Not viable for production use when null checking is common

**Additional context**
- This feature is critical for applications that need to frequently query for records with null values in specific fields
- The current workarounds either sacrifice performance (direct null comparison) or create data maintenance overhead (boolean conversion)
- The use case involves frequent queries where checking for null values is part of most query patterns, making performance crucial

**Comment by coszio** [comment]
Looking for this?

https://qdrant.tech/documentation/concepts/filtering/#is-null

Worth to note that any payload index should support this condition.

**Comment by esteininger** [comment]
> Looking for this?
> 
> https://qdrant.tech/documentation/concepts/filtering/#is-null

according to the docs it's not a listed field type for indexing:

https://qdrant.tech/documentation/concepts/indexing/#payload-index

fwiw null != false

**Comment by generall** [comment]
Issue make sense, I think we can create a special type of index, which will keep track of the empty and null values of index fields. It should not create overhead for those fields, which have no empty values.

**Comment by coszio** [comment]
Yes, documentation could be clearer regarding the indexing for this filter.

I have a question though, how were you doing this `eq null` comparison without `IsNull`?

**Comment by generall** [comment]
we will do it with `IsNull` but optimized
